Manage resource iterators explicitly

andreikoval opened the pull request:
Knowledge about iterators' inherent resources must not leak outside
relevant scope.  Responsibility of managing such resources also must be
explicit, not hidden as an implicit "instanceof" check in the totally
irrelevant part of code.

burqen reviewed (changes requested):
Changes look good! However, we add one extra layer if wrapping to a lot of index calls. I'll start a benchmark run to see how this affect performance. Hold tight.

burqen commented:
Benchmarks look fine, approved!

andreikoval commented:
As discussed offline, benchmark ui was confusing and there's actually a regression. Alternative solution will be implemented.

Commit:
Manage resource iterators explicitly

Knowledge about iterators' inherent resources must be reflected in
API contracts, not hidden under the hood as implicit "instanceof" check.

burqen commented on community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/ReadableRelationshipDiffSets.java:
```diff
@@ -128,4 +123,26 @@ public void accept( DiffSetsVisitor<T> visitor )
+
+    RelationshipIterator augment( RelationshipIterator source );
+
+    RelationshipIterator augmentWithRemovals( RelationshipIterator source );
```
This guy is never used

burqen commented on community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/ReadableRelationshipDiffSets.java:
```diff
@@ -24,13 +24,14 @@
  * {@link SuperReadableDiffSets} with added method for filtering added relationships.
  */
-public interface ReadableRelationshipDiffSets<T> extends SuperReadableDiffSets<T,RelationshipIterator, RelationshipIterator>
+public interface ReadableRelationshipDiffSets<T> extends SuperReadableDiffSets<T>
```
Can we place `Empty` implementation at the bottom and the interesting stuff (the methods) above?

burqen reviewed (approved):
This is so much cleaner, I really like this! Nice 👍
I'll approve this now, but please have a look at the small readability comments before merging.

burqen commented:
The test failure is not related to this PR. A rebase should fix it.

andreikoval commented:
Cool, thanks! I'll fix those readability issues.

Commit:
Removed unused method

Commit:
Minor readability fix
